### PR TITLE
Add 2.5 blog

### PR DESCRIPTION
**Pull request by sosukesuzuki**
## Description

https://deploy-preview-11823--prettier.netlify.app/blog/2021/11/25/2.5.0.html

## Checklist

- [x] I’ve read the [contributing guidelines](https://github.com/prettier/prettier/blob/main/CONTRIBUTING.md).

**✨[Try the playground for this PR](https://prettier.io/playground-redirect)✨**

### CODE REVIEW

File: website/blog/2021-11-17-2.5.0.md [hunk 1]
```diff
@@ -0,0 +1,337 @@
...
+```
+
+<!-- Prettier 2.4 -->
+```tsx
+const test = <T,>(value: T) => {};
+```
+
+<!-- Prettier 2.5 -->
+```tsx
+const test = <T>(value: T) => {};
```
Comment (fisker) [inline website/blog/2021-11-17-2.5.0.md:329]: Is this the output in 2.4?

File: changelog_unreleased/BLOG_POST_INTRO_TEMPLATE.md [hunk 2]
```diff
@@ -1,7 +1,9 @@
...
-author: "🚧"
-authorURL: "https://github.com/🚧"
-title: "Prettier 🚧"
+author: "Sosuke Suzuki (@sosukesuzuki)"
+authorURL: "https://github.com/sosukesuzuki"
+title: "Prettier 2.5: TypeScript 4.5 and MDX 2 comments!"
 ---
 
-🚧 Write an introduction here.
+This release adds support for [TypeScript 4.5](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-rc/) new syntax and [MDX 2](https://mdxjs.com/blog/v2/) comments!
```
Comment (kachkaev) [inline changelog_unreleased/BLOG_POST_INTRO_TEMPLATE.md:7]:
```suggestion
This release adds support for [TypeScript 4.5](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/) new syntax and [MDX 2](https://mdxjs.com/blog/v2/) comments!
```

File: website/blog/2021-11-18-2.5.0.md [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+// Prettier 2.5
+class Test {
+  @foo`bar`
+  test: string = "test"
+}
+```
+
+### SCSS
+
+#### Fix wildcard syntax in `@forward` ([#11482](https://github.com/prettier/prettier/pull/11482)) ([#11487](https://github.com/prettier/prettier/pull/11487) by [@niksy](https://github.com/niksy))
```
Comment (kachkaev) [inline website/blog/2021-11-18-2.5.0.md:239]: #11482 is an issue, not a pull request. How about removing this link from [`changelog_unreleased/scss/11487.md`](https://github.com/prettier/prettier/blob/59b5eb4e1faf912dd46759c98b45cfe5807e7a9f/changelog_unreleased/scss/11487.md) to make all header structures consistent?

```diff
- #### Fix wildcard syntax in `@forward` (#11482) (#11487 by @niksy)
+ #### Fix wildcard syntax in `@forward` (#11487 by @niksy)
```

Comment (kachkaev) [reply website/blog/2021-11-18-2.5.0.md:239]: Hmm, odd. Looks like this fix was already mentioned in [2.4.1 CHANGELOG](https://github.com/prettier/prettier/blob/main/CHANGELOG.md#241) 🤔

File: website/blog/2021-11-18-2.5.0.md [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+
+<!-- Prettier 2.5 -->
+```tsx
+const test = <T,>(value: T) => {};
+```
+````
+
+### CLI
+
+#### Add new CLI option `debug-print-ast` ([#11514](https://github.com/prettier/prettier/pull/11514) by [@sosukesuzuki](https://github.com/sosukesuzuki))
```
Comment (kachkaev) [inline website/blog/2021-11-18-2.5.0.md:364]: According to [CHANGELOG](https://github.com/prettier/prettier/blob/main/CHANGELOG.md#241), this feature has already been released in 2.4.1 🤔

File: website/blog/2021-11-18-2.5.0.md (hunk 3 above)
Comment (sosukesuzuki) [reply website/blog/2021-11-18-2.5.0.md:239]: Thank you. I forgot clean `changelog_unreleased` after releasing 2.4.1..

**Comment by kachkaev** [comment]
If anyone wants to proofread the post, note that Preview URL has changed:

~~https://deploy-preview-11823--prettier.netlify.app/blog/2021/11/17/2.5.0.html~~
https://deploy-preview-11823--prettier.netlify.app/blog/2021/11/18/2.5.0.html

File: website/blog/2021-11-18-2.5.0.md [hunk 5]
```diff
@@ -0,0 +1,366 @@
...
+Input:
+
+{/*A comment*/}
+
+Prettier 2.4:
+
+{/_A comment_/}
+
+Prettier 2.5:
+{/*A comment*/}
```
Comment (kachkaev) [inline website/blog/2021-11-18-2.5.0.md:110]:
```diff
 Input:
-
 {/*A comment*/}

 Prettier 2.4:
-
 {/_A comment_/}

 Prettier 2.5:
 {/*A comment*/}
```

Sorry for bikeshedding, but my brain could not walk past this 😅

File: changelog_unreleased/typescript/11515.md [hunk 6]
```diff
@@ -1,6 +1,7 @@
-#### Avoid extra offset in arrow function body when using long types (#11515 by @kachkaev and @thorn0)
+#### [HIGHLIGHT]Avoid extra offset in arrow function body when using long types (#11515 by @kachkaev and @thorn0)
 
 Starting with Prettier 2.3.0, type declarations in arrow functions could affect function body offset. Changing the length of the type annotation could produce large diffs and thus increased the chance of git conflicts. To prevent this, function body offset was stabilized.
+However, this change can make a big difference from 2.4.1. Please be careful.
```
Comment (j-f1) [inline changelog_unreleased/typescript/11515.md:4]: The use of the term “offset” here wasn’t clear to me — maybe change it to “indentation?” Additionally, I would reword the last sentence like so:

```suggestion
**Note**: This change may affect a large number of lines in your codebase.
```
(or something similar?)

File: changelog_unreleased/typescript/11721.md [hunk 7]
```diff
@@ -1,8 +1,8 @@
-#### Support TypeScript 4.5 (#11721, #11723, #11813 by @sosukesuzuki)
+#### [HIGHLIGHT]Support TypeScript 4.5 (#11721, #11723, #11813 by @sosukesuzuki)
 
 Support new TypeScript 4.5 features.
```
Comment (j-f1) [inline changelog_unreleased/typescript/11721.md:3]:
```suggestion
We’ve added support for new syntax features in TypeScript 4.5:
```

File: changelog_unreleased/typescript/11721.md [hunk 8]
```diff
@@ -17,7 +17,7 @@ import { type A } from "mod";
 
 ```
```
Comment (j-f1) [inline changelog_unreleased/typescript/11721.md:18]: Since these reflect new syntax, I think it would be better to provide examples rather than showing that there used to be a syntax error:

```ts
// Example
import { type A } from "mod";
```

File: changelog_unreleased/typescript/11721.md [hunk 9]
```diff
@@ -49,7 +49,7 @@ class Foo {
...
 
-##### [Import Assertions](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5-rc/#import-assertions)
+##### [Import Assertions](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/#import-assertions)
 
 <!-- prettier-ignore -->
 ```ts
```
Comment (j-f1) [inline changelog_unreleased/typescript/11721.md:55]: I can’t comment on the relevant line below, but I would word it as:

> Prettier will now parse files with `.mts` and `.cts` extensions as TypeScript.

(parse? see? handle? treat? not sure what’s best)

File: changelog_unreleased/mdx/11563.md [hunk 10]
```diff
@@ -1,21 +1,16 @@
-#### Add support for MDX 2 comments (#11563 by @wooorm)
+#### [HIGHLIGHT]Add support for MDX 2 comments (#11563 by @wooorm)
 
-This adds basic support for MDX 2 comments (JavaScript-style comments) in
-addition to existing MDX 1 comments (HTML-style comments).
+This adds basic support for MDX 2 comments (JavaScript-style comments) in addition to existing MDX 1 comments (HTML-style comments).
```
Comment (j-f1) [inline changelog_unreleased/mdx/11563.md:3]:
```suggestion
This adds basic support for MDX v2 comment syntax (JavaScript-style comments) in addition to the existing support MDX v1 comment syntax (HTML-style comments).
```

File: changelog_unreleased/mdx/11563.md [hunk 11]
```diff
@@ -1,21 +1,16 @@
...
 
-Note that in MDX 2, more complex comments and whole expressions can be used
-inside braces.
-This currently only supports a single comment on one line so that
-`{/* prettier-ignore */}` can be used, but doesn’t support most of MDX 2.
+Note that in MDX 2, more complex comments and whole expressions can be used inside braces.
+This currently only supports a single comment on one line so that `{/* prettier-ignore */}` can be used, but doesn’t support most of MDX 2.
```
Comment (j-f1) [inline changelog_unreleased/mdx/11563.md:6]:
```suggestion
**Note**: Prettier currently only supports the new comment syntax for single-line comments (so that `{/* prettier-ignore */}` can be used), and doesn’t support the rest of MDX v2.
```

File: website/blog/2021-11-21-2.5.0.md [hunk 12]
```diff
@@ -0,0 +1,384 @@
...
+{/*A comment*/}
+```
+
+## Other Changes
+
+### JavaScript
+
+#### Fix parentheses around sequence expression as body of arrow chain ([#11593](https://github.com/prettier/prettier/pull/11593) by [@bakkot](https://github.com/bakkot))
+
+The required parentheses around sequence expressions as the body of arrow functions were dropped for chained arrows, causing incorrect code to be generated. Now they are included.
```
Comment (j-f1) [inline website/blog/2021-11-21-2.5.0.md:155]:
```suggestion
The required parentheses around sequence expressions as the body of arrow functions are now preserved for chained arrows. Previously, Prettier removed them, which resulted in invalid syntax.
```

File: website/blog/2021-11-21-2.5.0.md [hunk 13]
```diff
@@ -0,0 +1,384 @@
...
+
+// Prettier 2.4
+const f = () => () => 0, 1;
+
+// Prettier 2.5
+const f = () => () => (0, 1);
+```
+
+#### Ignore errors for sloppy mode syntax ([#11750](https://github.com/prettier/prettier/pull/11750) by [@fisker](https://github.com/fisker), [#11778](https://github.com/prettier/prettier/pull/11778) by [@sosukesuzuki](https://github.com/sosukesuzuki))
+
```
Comment (j-f1) [inline website/blog/2021-11-21-2.5.0.md:170]:
```suggestion

JavaScript’s [strict mode](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Strict_mode) adds several useful errors to prevent mistakes. Some of these errors are syntax errors that occur at parse time. Since Prettier’s goal is to format all syntactically valid JavaScript code regardless of whether it will actually run, we’ve opted to leave this error checking to linters, compilers, and the runtime.

```

File: website/blog/2021-11-21-2.5.0.md [hunk 14]
```diff
@@ -0,0 +1,384 @@
...
+const paragraph2 = css`
+  transform: ${expr} (30px);
+`;
+
+// Prettier 2.5
+const paragraph2 = css`
+  transform: ${expr}(30px);
+`;
+
+```
```
Comment (j-f1) [inline website/blog/2021-11-21-2.5.0.md:209]:
```suggestion
<!-- prettier-ignore -->
```jsx
// Input
const paragraph2 = css`
  transform1: ${expr}(30px);
  transform2: ${expr} (30px);
`;

// Prettier 2.4
const paragraph2 = css`
  transform1: ${expr} (30px);
  transform2: ${expr} (30px);
`;

// Prettier 2.5
const paragraph2 = css`
  transform1: ${expr}(30px);
  transform2: ${expr} (30px);
`;

```
```

File: website/blog/2021-11-21-2.5.0.md [hunk 15]
```diff
@@ -0,0 +1,384 @@
...
+  6 | }
+
+// Prettier 2.5
+class Foo {
+  #brand;
+  static isC(obj) {
+    return #brand in Foo;
+  }
+}
+```
```
Comment (j-f1) [inline website/blog/2021-11-21-2.5.0.md:239]: Similarly to the TypeScript thing I mentioned earlier:

```js
// Example
class Foo {
  #brand;
  static isC(obj) {
    return #brand in Foo;
  }
}
```

File: website/blog/2021-11-21-2.5.0.md [hunk 16]
```diff
@@ -0,0 +1,384 @@
...
+  test: string = "test"
+
+  @test().x("global").y()
+  test2: string = "test";
+}
+```
+
+### SCSS
+
+#### Improve `with (...)` formatting ([#11637](https://github.com/prettier/prettier/pull/11637) by [@sosukesuzuki](https://github.com/sosukesuzuki))
```
Comment (j-f1) [inline website/blog/2021-11-21-2.5.0.md:277]: For consistency with the next entry:

```suggestion
#### Improve `@use with` formatting ([#11637](https://github.com/prettier/prettier/pull/11637) by [@sosukesuzuki](https://github.com/sosukesuzuki))
```

File: changelog_unreleased/html/11827.md [hunk 17]
```diff
@@ -1 +1,26 @@
 #### Reverts #7865, keeping class names on the same line (#11827 by @jlongster)
```
Comment (j-f1) [inline changelog_unreleased/html/11827.md:1]: I think this should be a highlight since it is likely to substantially affect some users’ code.

```suggestion
#### Collapse HTML `class` attributes onto one line (#11827 by @jlongster, reverts #7865)
```

Also, a change like this should probably have a rationale. Here’s my proposal, but as I’m not an active maintainer, I think y’all’s opinions are more valuable.

> While this was intended to be useful for users of CSS libraries like Tailwind that tend to result in large numbers of classes on elements, it became clear that our heuristics for where to split the class list on to multiple lines were unable to consistently produce good results. We’re still considering better ways to format HTML with lots of classes — consider `[discussing with us](https://github.com/prettier/prettier/issues/7863)`.

File: website/blog/2021-11-21-2.5.0.md [hunk 18]
```diff
@@ -0,0 +1,384 @@
...
+{{!-- Prettier 2.4 --}}
+<div title="{{t "login.username.description"}}" />
+
+```
+
+### Markdown
+
+#### Keep trailing commas for type parameters in embedded TSX ([#11685](https://github.com/prettier/prettier/pull/11685) by [@sosukesuzuki](https://github.com/sosukesuzuki))
+
+<!-- Optional description if it makes sense. -->
```
Comment (j-f1) [inline website/blog/2021-11-21-2.5.0.md:366]:
```suggestion
The trailing comma is necessary to prevent TypeScript from treating the `<T>` as the beginning of a JSX expression. 
```

**Review by j-f1 — COMMENTED** [review]
Made a bunch of suggestions. I’m not particularly attached to them, so feel free to reject any you don’t agree with :)

File: changelog_unreleased/typescript/11721.md (hunk 8 above)
Comment (sosukesuzuki) [reply changelog_unreleased/typescript/11721.md:18]: Sounds good!

File: changelog_unreleased/typescript/11721.md (hunk 9 above)
Comment (sosukesuzuki) [reply changelog_unreleased/typescript/11721.md:55]: format?

Comment (j-f1) [reply changelog_unreleased/typescript/11721.md:31]: That works!

File: changelog_unreleased/BLOG_POST_INTRO_TEMPLATE.md [hunk 19]
```diff
@@ -1,7 +1,9 @@
 ---
-author: "🚧"
-authorURL: "https://github.com/🚧"
-title: "Prettier 🚧"
+author: "Sosuke Suzuki (@sosukesuzuki)"
```
Comment (azz) [inline changelog_unreleased/BLOG_POST_INTRO_TEMPLATE.md:2]: Did you mean to update the template here?

File: changelog_unreleased/blog-post-intro.md [hunk 20]
```diff
@@ -0,0 +1,9 @@
+---
+author: "Sosuke Suzuki (@sosukesuzuki)"
+authorURL: "https://github.com/sosukesuzuki"
+title: "Prettier 2.5: TypeScript 4.5 and MDX v2 comment syntax!"
+---
+
+This release adds support for [TypeScript 4.5](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/) new syntax and [MDX v2](https://mdxjs.com/blog/v2/) comment syntax!
```
Comment (azz) [inline changelog_unreleased/blog-post-intro.md:7]:
```suggestion
This release adds support for [TypeScript 4.5](https://devblogs.microsoft.com/typescript/announcing-typescript-4-5/)'s new syntax and [MDX v2](https://mdxjs.com/blog/v2/) comment syntax!
```

File: changelog_unreleased/typescript/11721.md [hunk 21]
```diff
@@ -1,45 +1,21 @@
-#### Support TypeScript 4.5 (#11721, #11723, #11813 by @sosukesuzuki)
+#### [HIGHLIGHT]Support TypeScript 4.5 (#11721, #11723, #11813 by @sosukesuzuki)
 
-Support new TypeScript 4.5 features.
+We’ve added support for new syntax features in TypeScript 4.5:
```
Comment (azz) [inline changelog_unreleased/typescript/11721.md:3]: Little bit ambiguous.

```suggestion
We’ve added support for TypeScript 4.5's new syntax features:
```

File: changelog_unreleased/typescript/11515.md [hunk 22]
```diff
@@ -1,6 +1,7 @@
-#### Avoid extra offset in arrow function body when using long types (#11515 by @kachkaev and @thorn0)
+#### [HIGHLIGHT]Avoid extra offset in arrow function body when using long types (#11515 by @kachkaev and @thorn0)
 
 Starting with Prettier 2.3.0, type declarations in arrow functions could affect function body offset. Changing the length of the type annotation could produce large diffs and thus increased the chance of git conflicts. To prevent this, function body offset was stabilized.
+**Note**: This change may affect a large number of lines in your codebase.
```
Comment (j-f1) [inline changelog_unreleased/typescript/11515.md:4]:
```suggestion
Starting with Prettier 2.3.0, type declarations in arrow functions could affect function body offset. Changing the length of the type annotation could produce large diffs and thus increased the chance of git conflicts. To prevent this, function body offset was stabilized.<br>
**Note**: This change may affect a large number of lines in your codebase.
```

**Review by j-f1 — APPROVED** [review]
LGTM with one small formatting suggestion:

**Review by fisker — APPROVED** [review]

**Review by kachkaev — APPROVED** [review]

File: website/blog/2021-11-24-2.5.0.md [hunk 23]
```diff
@@ -0,0 +1,341 @@
...
+  MyComponentWithLongNameProps
+> = ({ x, y }) => {
+  const a = useA();
+  return <div>{x + y + a}</div>;
+};
+```
+
+#### Support TypeScript 4.5 ([#11721](https://github.com/prettier/prettier/pull/11721), [#11723](https://github.com/prettier/prettier/pull/11723), [#11813](https://github.com/prettier/prettier/pull/11813) by [@sosukesuzuki](https://github.com/sosukesuzuki))
+
+We’ve added support for TypeScript 4.5's new syntax features:
```
Comment (kachkaev) [inline website/blog/2021-11-24-2.5.0.md:57]: A veeeeery minor thing that’s not even worth a push if there is nothing else to change. I’ve noticed the usage of `'` (typewriter apostrophe) instead of `’` ([typographic apostrophe](https://en.wikipedia.org/wiki/Apostrophe#Typographic_form)) in a couple of places. Could be worth searching for `'s` and replacing with`’s` for consistency with most text. Note that `'` is used in code snippets so we should be careful 😁

**Comment by j-f1** [comment]
I just noticed that https://github.com/prettier/prettier/pull/11823#discussion_r753836150 was marked as resolved but my first sentence about the use of “offset” was not addressed.

**Comment by sosukesuzuki** [comment]
@j-f1 @kachkaev Thank you for your reviewing! I've fixed.

File: website/blog/2021-11-24-2.5.0.md [hunk 24]
```diff
@@ -0,0 +1,341 @@
...
+### SCSS
+
+#### Improve `@use with` formatting ([#11637](https://github.com/prettier/prettier/pull/11637) by [@sosukesuzuki](https://github.com/sosukesuzuki))
+
+<!-- prettier-ignore -->
+```scss
+// Input
+@use 'library' with (
+  $black: #222,
+  $border-radius: 0.1rem $font-family: 'Helvetica, sans-serif'
```
Comment (glen-84) [inline website/blog/2021-11-24-2.5.0.md:270]: Is this supposed to be invalid Sass? If not, there's a missing comma. (below as well + newline)

**Review by lipis — APPROVED** [review]

File: website/blog/2021-11-24-2.5.0.md (hunk 24 above)
Comment (sosukesuzuki) [reply website/blog/2021-11-24-2.5.0.md:270]: I think it is valid scss

Comment (glen-84) [reply website/blog/2021-11-24-2.5.0.md]: How?

Comment (sosukesuzuki) [reply website/blog/2021-11-24-2.5.0.md]: I'm not familiar with SCSS, but at least there is a below code on the scss docs.

```scss
@use 'library' with (
  $black: #222,
  $border-radius: 0.1rem
);
```

https://sass-lang.com/documentation/at-rules/use#configuration

Comment (glen-84) [reply website/blog/2021-11-24-2.5.0.md]: image

There are two properties on the same line, with no comma between them.

Comment (sosukesuzuki) [reply website/blog/2021-11-24-2.5.0.md]: Woops sorry I didn't notice that. I'll fix later, thank you!

**Comment by kachkaev** [comment]
Great work @sosukesuzuki 👏👏👏
